packages/ava/src/text/generate/generateTextSpec.ts

```diff
@@ -0,0 +1,128 @@
+import { isUndefined, arr2map } from '../../utils';
+
+import type { NarrativeTextSpec, ParagraphSpec, PhraseSpec, TextParagraphSpec, TextPhraseSpec } from '../schema';
+import type { Structure, Variable, StructureTemp } from './types';
+
+function getSentences(
+  template: string,
+  dataValue?: Variable['dataValue'],
+  dataMetaMap?: Variable['dataMetaMap']
+): PhraseSpec[][] {
+  const splitReg = /(\${.*?})/;
+  const varReg = /\${(.*?)}/;
+  const METRIC_NAME_SUFFIX = '[metric_name]';
+
+  // "user ${name}, age is ${age}" => ["user ", "${name}", ", age is ", "${age}"]
+  const templateStrArr = template.split(splitReg).filter((str) => str);
+  // {} => [{}]
+  const formattedDataValue = Array.isArray(dataValue) ? dataValue : [dataValue || {}];
+
+  return formattedDataValue.map((datum) => {
+    return templateStrArr.map<PhraseSpec>((str) => {
+      const varName: string | undefined = varReg.exec(str)?.[1]; // age
+      if (varName) {
+        if (varName.endsWith(METRIC_NAME_SUFFIX)) {
+          const fieldId = varName.slice(0, -METRIC_NAME_SUFFIX.length);
+          return {
+            type: 'entity',
+            value: dataMetaMap?.[fieldId]?.name || fieldId,
+            metadata: { entityType: 'metric_name' },
+          };
+        }
+
+        const data = isUndefined(datum[varName]) ? '-' : `${datum[varName]}`;
+        const dataMeta = dataMetaMap?.[varName];
+        if (dataMeta?.entityType) {
+          return { type: 'entity', value: data, metadata: { entityType: dataMeta?.entityType } };
+        }
+        return { type: 'text', value: data };
+      }
+      return { type: 'text', value: str };
+    });
+  });
+}
+
+/**
+ * use structure and variables to generate narrative text spec
+ */
+function generateTextSpec({
+  structures,
+  structureTemps = [],
+  variables = [],
+}: {
+  structures: Structure[];
+  structureTemps?: StructureTemp[];
+  variables?: Variable[];
+}): NarrativeTextSpec {
+  const variableMap = arr2map<Variable>(variables, 'variableId');
+  const structureTempMap = arr2map<StructureTemp>(structureTemps, 'templateId');
+  const splitReg = /(&{.*?})/;
+  const varReg = /&{(.*?)}/;
+
+  // generate paragraphs by structure
+  let paragraphs: ParagraphSpec[] = [];
```

Comment:
Do u want to reset it? If so, u could `arr.splice(0, arr.length, newValue)`
If not, u could write `push([...])` directly.

I think using `const` could make everybody know it's constant, and it will bring confidence, make us using it without worrying about its changes. Even if it is an array or object, we could know its address remains unchanged.